### PR TITLE
feat(gear-programs): Erc20 relay program

### LINKED ISSUE #74: Ethereum light client gear programs

### PR DESCRIPTION
**Pull request by gshep**
Resolves #74

### CODE REVIEW

File: gear-programs/erc20-relay/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "erc20-relay"
+version = "0.1.0"
+edition = "2021"
```
Comment (mertwole) [inline gear-programs/erc20-relay/Cargo.toml:4]:
```suggestion
version.workspace = true
edition.workspace = true

```

File: gear-programs/erc20-relay/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+erc20-relay-app.workspace = true
+
+[build-dependencies]
+erc20-relay-app.workspace = true
+sails-rs = { git = "https://github.com/gear-tech/sails.git", rev = "aab08733ee9ccdb809dc9b29c57dd411b2b917b1", features = ["wasm-builder"] }
+sails-idl-gen = { git = "https://github.com/gear-tech/sails.git", rev = "aab08733ee9ccdb809dc9b29c57dd411b2b917b1" }
```
Comment (mertwole) [inline gear-programs/erc20-relay/Cargo.toml:12]: If you plan to merge it without waiting for `sails` release then please open an issue and link it here

File: gear-programs/erc20-relay/app/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "erc20-relay-app"
+version = "0.1.0"
+edition = "2021"
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/Cargo.toml:4]:
```suggestion
version.workspace = true
edition.workspace = true
```

File: gear-programs/erc20-relay/Cargo.toml (hunk 2 above)
Comment (mertwole) [reply gear-programs/erc20-relay/Cargo.toml:12]: And can it be moved to sworkspace deps to not forget to update it everywhere in future?

File: gear-programs/erc20-relay/app/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+[dependencies]
+sails-rs = { git = "https://github.com/gear-tech/sails.git", rev = "aab08733ee9ccdb809dc9b29c57dd411b2b917b1" }
+ethereum-common.workspace = true
+alloy-rlp.workspace = true
+alloy-sol-types = { workspace = true, features = ["json"] }
+checkpoint_light_client-io.workspace = true
+
+[target.'cfg(all(target_arch = "wasm32", target_os = "unknown"))'.dependencies]
+getrandom = { version = "0.2", default-features = false, features = ["custom"] }
+lazy_static = { version = "1.1", features = ["spin_no_std"] }
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/Cargo.toml:15]: workspace deps

File: gear-programs/erc20-relay/app/src/ERC20Treasury.json
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/ERC20Treasury.json:1]: You can bind to it in `ethereum/out/ERC20Treasury.sol/ERC20Treasury.json`

File: gear-programs/erc20-relay/app/build.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+use sails_client_gen::ClientGenerator;
+use std::{env, path::PathBuf};
+
+fn main() {
+    let mut path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+    path.pop();
+    path.pop();
+
+    let idl_file_path = path.join("vft-gateway/src/wasm/vft-gateway.idl");
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/build.rs:9]: Will this path work on windows?

File: gear-programs/erc20-relay/app/src/services/mod.rs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+use super::*;
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/services/mod.rs:1]:
```suggestion
```

File: gear-programs/erc20-relay/client/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "erc20-relay-client"
+version = "0.1.0"
+edition = "2021"
```
Comment (mertwole) [inline gear-programs/erc20-relay/client/Cargo.toml:4]:
```suggestion
version.workspace = true
edition.workspace = true
```

File: gear-programs/erc20-relay/client/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+[dependencies]
+mockall = { version = "0.12", optional = true }
```
Comment (mertwole) [inline gear-programs/erc20-relay/client/Cargo.toml:7]: workspace deps

File: gear-programs/erc20-relay/src/lib.rs
Comment (mertwole) [inline gear-programs/erc20-relay/src/lib.rs:1]: Why the structure is different from one that's used commonly with `sails`?

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+            panic!("Not admin");
+        }
+
+        state.vft_gateway = vft_gateway;
+    }
+
+    pub async fn relay(&mut self, message: EthToVaraEvent) -> Result<(), Error> {
+        let Some(vft_gateway_id) = self.state.borrow().vft_gateway else {
+            return Err(Error::AbsentVftGateway);
+        };
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/services/erc20_relay.rs:57]: Why not to set it in `init` then? to not deal with `Option`

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs [hunk 10]
```diff
@@ -0,0 +1,274 @@
+// Incorporate code generated based on the IDL file
+#[allow(dead_code)]
+mod vft {
+    include!(concat!(env!("OUT_DIR"), "/vft-gateway.rs"));
+}
+
+use super::*;
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/services/erc20_relay.rs:7]: Please don't use `*` imports enywhere. It complicates code-review and navigation in source files. For example, if I want to see where the `State` is declared I need to go to `super` crate (and if there're some re-exports in `super` I also need to search among all of them)

File: gear-programs/erc20-relay/app/src/services/ft_manage.rs
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/services/ft_manage.rs:1]: It's better to change `vft-gateway::mint` signature to accept ethereum token id as it already stores such a mapping and we don't need to maintain both

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        let _ = self.notify_on(Event::Relayed {
+            fungible_token,
+            to: ActorId::from(event.to.0),
+            amount,
+        });
+
+        Ok(())
+    }
+
+    fn prepare(
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/services/erc20_relay.rs:137]: It's not very clear from method name what are we preparing. Also struct for return type will be great

File: gear-programs/erc20-relay/src/lib.rs
Comment (gshep) [reply gear-programs/erc20-relay/src/lib.rs:1]: Hm it is one generated by `cargo sails new-program ...`. I only edited dependencies set to make them workspace deps.

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs (hunk 9 above)
Comment (gshep) [reply gear-programs/erc20-relay/app/src/services/erc20_relay.rs:57]: Yep, it is possible. However then we need to implement such logic in `vft-gateway` program. So I decided to make it here

File: gear-programs/erc20-relay/app/src/services/ft_manage.rs
Comment (gshep) [reply gear-programs/erc20-relay/app/src/services/ft_manage.rs:1]: Done in #133

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs (hunk 10 above)
Comment (gshep) [reply gear-programs/erc20-relay/app/src/services/erc20_relay.rs:7]: However I wonder why it makes issues. It is common pattern, for example `use sails_rs::{cell::RefCell, prelude::*};` - https://github.com/gear-tech/sails/blob/master/examples/demo/app/src/lib.rs#L4

Anyway I will edit it if you insist ).

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs (hunk 9 above)
Comment (gshep) [reply gear-programs/erc20-relay/app/src/services/erc20_relay.rs:57]: I mean that both programs must have each other's addresses when initialized. Unfortunately, this cannot be done without Option .

However, the question remains, where to place such logic: here in erc 20-relay-app or in vft-gateway?

File: gear-programs/erc20-relay/app/src/ERC20Treasury.json
Comment (gshep) [reply gear-programs/erc20-relay/app/src/ERC20Treasury.json:1]: Looks like it's not quite correct: we have to use the JSON-file with an ABI-description generated for the deployed contract.

Comment (mertwole) [reply gear-programs/erc20-relay/app/src/ERC20Treasury.json:1]: discussed. Bind to ethereum/out/...

File: gear-programs/erc20-relay/src/lib.rs
Comment (mertwole) [reply gear-programs/erc20-relay/src/lib.rs:1]: discussed in dm

File: gear-programs/erc20-relay/app/src/services/erc20_relay.rs (hunk 9 above)
Comment (mertwole) [reply gear-programs/erc20-relay/app/src/services/erc20_relay.rs:62]: discussed: use Dfeault::default entity

File: gear-programs/erc20-relay/app/src/error.rs [hunk 12]
```diff
@@ -0,0 +1,21 @@
+use super::*;
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/error.rs:1]:
```suggestion
```

File: gear-programs/erc20-relay/app/src/service.rs [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+            return Err(Error::TooOldTransaction);
+        }
+
+        Ok(CheckedReceipt {
+            envelope: receipt,
+            event,
+        })
+    }
+
+    pub async fn request_checkpoint(checkpoints: ActorId, slot: u64) -> Result<H256, Error> {
```
Comment (mertwole) [inline gear-programs/erc20-relay/app/src/service.rs:240]:
```suggestion
    async fn request_checkpoint(checkpoints: ActorId, slot: u64) -> Result<H256, Error> {
```

**Review by mertwole — APPROVED** [review]

**Pull request merged**